validator/server_arb/execution_run.go

```diff
@@ -79,6 +83,119 @@ func (e *executionRun) GetStepAt(position uint64) containers.PromiseInterface[*v
 	})
 }
 
+func (e *executionRun) GetMachineHashesWithStepSize(fromBatch, machineStartIndex, stepSize, numDesiredLeaves uint64) containers.PromiseInterface[[]common.Hash] {
+	return stopwaiter.LaunchPromiseThread(e, func(ctx context.Context) ([]common.Hash, error) {
+		return e.machineHashesWithStepSize(ctx, machineHashesWithStepSizeArgs{
+			startIndex:        machineStartIndex,
+			fromBatch:         fromBatch,
+			stepSize:          stepSize,
+			requiredNumHashes: numDesiredLeaves,
+		})
+	})
+}
+
+type GlobalStateGetter interface {
+	GetGlobalState() validator.GoGlobalState
+	HashStepper
+}
+
+type HashStepper interface {
+	Step(ctx context.Context, stepCount uint64) error
+	Hash() common.Hash
+}
+
+type machineHashesWithStepSizeArgs struct {
+	startIndex        uint64
+	fromBatch         uint64
+	stepSize          uint64
+	requiredNumHashes uint64
+	getMachineAtIndex func(context.Context, uint64) (GlobalStateGetter, error)
+}
+
+func (e *executionRun) machineHashesWithStepSize(
```

Comment:
If this function is a member of executionRun, it shouldn't get "getMachineAtIndex" as parameter because that should always be from the executionRun itself.